libdrgn/arch_aarch64.c

```diff
@@ -409,78 +399,70 @@ linux_kernel_pgtable_iterator_next_aarch64(struct drgn_program *prog,
 					   uint64_t *virt_addr_ret,
 					   uint64_t *phys_addr_ret)
 {
-	struct drgn_error *err;
 	const uint64_t page_shift = prog->vmcoreinfo.page_shift;
 	const uint64_t pgtable_shift = page_shift - 3;
 	const bool bswap = drgn_platform_bswap(&prog->platform);
 	struct pgtable_iterator_aarch64 *it =
 		container_of(_it, struct pgtable_iterator_aarch64, it);
 	const uint64_t virt_addr = it->it.virt_addr;
 	int level;
-
-	// Find the lowest level with cached entries.
-	for (level = 0; level < it->levels - 1; level++) {
-		if (it->index[level] < it->entries_per_level)
-			break;
+	uint16_t num_entries = it->last_level_num_entries;
+	uint64_t table = it->it.pgtable;
+	bool table_physical = false;
+
+	if (virt_addr < it->va_range_min ||
+	    virt_addr > it->va_range_max) {
+		*virt_addr_ret = it->va_range_min;
+		*phys_addr_ret = UINT64_MAX;
+		it->it.virt_addr = it->va_range_max + 1;
+		return NULL;
 	}
-	if (level == it->levels - 1 &&
-	    it->index[level] >= it->last_level_num_entries)
-		level++;
-	// For every level below that, refill the cache/return pages.
-	for (;; level--) {
-		uint16_t num_entries;
-		uint64_t table;
-		bool table_physical;
-		if (level == it->levels) {
-			num_entries = it->last_level_num_entries;
-			if (virt_addr < it->va_range_min ||
-			    virt_addr > it->va_range_max) {
-				*virt_addr_ret = it->va_range_min;
-				*phys_addr_ret = UINT64_MAX;
-				it->it.virt_addr = it->va_range_max + 1;
-				return NULL;
+
+	for (level = it->levels;; level--) {
+		uint8_t level_shift = page_shift + pgtable_shift * (level - 1);
+		uint16_t index = (virt_addr >> level_shift) & (num_entries - 1);
+		uint16_t cached_index = (it->cached_virt_addr >> level_shift) &
+					(num_entries - 1);
+		if (index != cached_index)
+			memset(it->table, 0, 8 * level);
+		uint64_t *entry_ptr = &it->table[level - 1];
+		if (!*entry_ptr) {
+			struct drgn_error *err = drgn_program_read_memory(
+			    prog, entry_ptr, table + 8 * index, 8,
+			    table_physical);
+			if (err) {
+				it->cached_virt_addr = 0;
+				memset(it->table, 0, sizeof(it->table));
+				return err;
 			}
```

Comment:
A page table iterator shouldn't be used after it returns an error (and none of the other architectures try to reinitialize after an error), so I'd prefer to get rid of this.

```suggestion
			if (err)
				return err;
```

Reply:
Done